src/SignalR/common/Http.Connections/src/Internal/HttpConnectionManager.cs

```diff
@@ -178,8 +182,14 @@ public void Scan()
             }
         }
 
-        public void CloseConnections()
+        public async Task CloseConnections()
         {
+            if (ShutdownDelay != TimeSpan.Zero)
+            {
+                // Delay to let users react to application shutdown before we close the SignalR connections
+                await Task.Delay(ShutdownDelay);
```

Comment:
Instead of adding a `TimeSpan? ShutdownDelay` option, can we add a `Func<Task> ShutdownCallback` option? The implementation could just be a `Task.Delay`, but it could also be more interesting.

Reply:
I don't think we should delay at all by default. That slows down the graceful shutdown of the server.

Reply:
I think we should have a default delay, the same way graceful shutdown delays before aggressively closing connections. Think of this similar to the logic we have in Kestrel. We need to give users the chance to gracefully shutdown their streams

Reply:
> I think we should have a default delay, the same way graceful shutdown delays before aggressively closing connections.

The difference is that Kestrel will shutdown immediately if there are no active requests. With this design, even the app developer proactively closes all their connections as soon as ApplicationStopping fires, this will *still* delay shutdown by an extra second. 😢

Reply:
Here's one thing I think we can do to get the best of both worlds. We need to be the last application stopping event in the pipeline so that applications can block stopping before we do our work. That would allow us to shutdown immediately and have customers code be able to do the sleep/wait/processing

Reply:
> With this design, even the app developer proactively closes all their connections as soon as ApplicationStopping fires, this will still delay shutdown by an extra second. 😢

I just realized I was wrong about this because of `_ = CloseConnections()`. Why are we even building this list of Tasks then? 😆 

> We need to be the last application stopping event in the pipeline so that applications can block stopping before we do our work.

@Tratcher had a comment about this when I suggested something similar in triage.

Reply:
> Here's one thing I think we can do to get the best of both worlds. We need to be the last application stopping event in the pipeline so that applications can block stopping before we do our work.

How do you guarantee that ordering?

Reply:
I don't think that we can guarantee ordering but I think we should start from that premise and work backwards. If we don't like the timeout idea, when we should think about ways to make it possible to achieve what people want. They want to write code that handles the shutdown before our logic kicks in. Maybe we need a new event...